commons/src/main/java/io/aiven/kafka/tieredstorage/commons/SegmentManifestProvider.java

```diff
@@ -0,0 +1,116 @@
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+import org.apache.kafka.server.log.remote.storage.RemoteLogSegmentMetadata;
+
+import io.aiven.kafka.tieredstorage.commons.manifest.SegmentManifest;
+import io.aiven.kafka.tieredstorage.commons.storage.FileFetcher;
+import io.aiven.kafka.tieredstorage.commons.storage.StorageBackEndException;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.benmanes.caffeine.cache.AsyncCacheLoader;
+import com.github.benmanes.caffeine.cache.AsyncLoadingCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+class SegmentManifestProvider {
+    private static final long GET_TIMEOUT_SEC = 10;
+
+    private final ObjectKey objectKey;
+    private final AsyncLoadingCache<String, SegmentManifest> cache;
+
+    /**
+     * @param maxCacheSize the max cache size (in items) or empty if the cache is unbounded.
+     * @param cacheRetention the retention time of items in the cache or empty if infinite retention.
+     */
+    SegmentManifestProvider(final ObjectKey objectKey,
+                            final Optional<Long> maxCacheSize,
+                            final Optional<Duration> cacheRetention,
+                            final FileFetcher fileFetcher,
+                            final ObjectMapper mapper,
+                            final Executor executor) {
+        this.objectKey = objectKey;
+        final var cacheBuilder = Caffeine.newBuilder()
+            .executor(executor);
+        maxCacheSize.ifPresent(cacheBuilder::maximumSize);
+        cacheRetention.ifPresent(cacheBuilder::expireAfterWrite);
+        this.cache = cacheBuilder.buildAsync(new Loader(fileFetcher, mapper));
+    }
+
+    private static class Loader implements AsyncCacheLoader<String, SegmentManifest> {
+        private final FileFetcher fileFetcher;
+        private final ObjectMapper mapper;
+
+        private Loader(final FileFetcher fileFetcher, final ObjectMapper mapper) {
+            this.fileFetcher = fileFetcher;
+            this.mapper = mapper;
+        }
+
+        @Override
+        public CompletableFuture<? extends SegmentManifest> asyncLoad(
+            final String key, final Executor executor) throws Exception {
+            final Supplier<SegmentManifest> fetcher = () -> {
+                try (final InputStream is = fileFetcher.fetch(key)) {
+                    return mapper.readValue(is, SegmentManifest.class);
+                } catch (final StorageBackEndException | IOException e) {
+                    // Wrap these exceptions to unwrap them later.
+                    throw new CompletionException(e);
+                }
+            };
+            return CompletableFuture.supplyAsync(fetcher, executor);
+        }
```

Comment:
fyi, I think you could have used `CacheLoader.load(key)` and let's default `asyncLoad` wrap that with a `supplyAsync`. A small quality-of-life nit.

Reply:
I think it would just be the following, assuming might prefer it inlined as a lambda
```java
this.cache = cacheBuilder.buildAsync(key -> {
  try (final InputStream is = fileFetcher.fetch(key)) {
    return mapper.readValue(is, SegmentManifest.class);
  }
}
```

Reply:
Thanks a lot @ben-manes, looks interesting and I guess will will consider this :+1: